backend/core/utils/tools.py

```
@@ -0,0 +1,162 @@
+            },
+        }
+
+    def _calculator(self, expression: str) -> str:
+        """Simple calculator with safety."""
+        try:
+            # Basic safety check: only allow numbers, operators, and whitespace
+            if re.search(r"[a-zA-Z]", expression):
+                return "Error: Unsafe expression. Only numbers and operators are allowed."
+            
+            # Using eval with restricted globals/locals for basic sandboxing
+            result = eval(expression, {"__builtins__": {}}, {})
```

Ruff (0.15.9)

[error] Use of possibly insecure function; consider using `ast.literal_eval`

(S307)